DM-24807: Add type annotations to datastore

timj opened the pull request.

timj commented on python/lsst/daf/butler/core/storedFileInfo.py:
```diff
@@ -45,8 +47,9 @@ class StoredFileInfo(StoredDatastoreItemInfo):
 
-    formatter: str
-    """Fully-qualified name of Formatter."""
+    formatter: Union[str, Formatter, Type[Formatter]]
```
This is quite problematic since it stops us from saying that we accept these three types but it will always be stored internally as a str.

TallJimbo replied:
Looks like there's no particularly good answer (https://stackoverflow.com/questions/51737828/python-dataclasses-what-type-to-use-if-post-init-performs-type-conversion doesn't provide one, for example).

I would probably lean towards just defining `__init__` directly instead of `__post_init__` so the parameter type can differ from the attribute type.  That's how I was planning to handle the `TableSpec` problem, which is essentially identical to this one.  But of course that's not ideal either.

timj replied:
I think we had a reason for using `__post_init__` since it seemed designed for this and didn't result in me trying to ensure that I implemented `__init__` for the dataclass properly.

timj replied:
I've rewritten it and things seem to work okay now.

TallJimbo commented on python/lsst/daf/butler/core/datasets/ref.py:
```diff
@@ -311,7 +325,8 @@ def _lookupNames(self) -> Tuple[LookupKey]:
     @staticmethod
-    def flatten(refs: Iterable[DatasetRef], *, parents: bool = True) -> Iterator[DatasetRef]:
+    def flatten(refs: Iterable[Union[FakeDatasetRef, DatasetRef]], *,
+                parents: bool = True) -> Iterator[DatasetRef]:
```
I don't think this one is the right way to fix whatever problem inspired it - if it's given a `FakeDatasetRef`, there's no way it's going to return only true `DatasetRefs`.  If the problem is that this is used to implement `FakeDatasetRef.flatten`, we should just `# type: ignore` that implementation.  There are probably other solutions, too, but they're more involved and that one seems pretty pragmatic.

timj replied:
Yes, this is an implementation detail in that I didn't want to duplicate the code from DatasetRef. I'm starting to wonder if inheriting from DatasetRef and then stubbing out all the methods that we know won't work wouldn't be easier.

timj replied:
I fixed it by adding a `allRefs` method so that mypy no longer gets confused abut the return type of the static method.

TallJimbo commented on python/lsst/daf/butler/core/datastore.py:
```diff
@@ -277,23 +288,23 @@ def __init__(self, config, bridgeManager, butlerRoot=None):
 
     @property
-    def names(self):
+    def names(self) -> List[str]:
```
Returning a `list` via a read-only property might be confusing - it's possible to write something like `obj.names[0] = "zero"`, and have it silently do nothing.  Could you return a `tuple` here instead?

timj replied:
That is a good point. I should be returning a tuple.

TallJimbo commented on python/lsst/daf/butler/core/datastore.py:
```diff
@@ -277,23 +288,23 @@ def __init__(self, config, bridgeManager, butlerRoot=None):
 
     @contextlib.contextmanager
-    def transaction(self):
+    def transaction(self) -> Generator:
```
I think you can give MyPy a bit more to work with by saying `Generator[None]`.

If that's the syntax for saying it yields `None`, that is - I'm not sure for `Generator`.  I've tended to use `Iterator` instead of `Generator` for functions decorated by `contextmanager`, because `Generator` feels like an implementation choice ather than an interface.  But `Generator` was put in `typing` by smarter people than me, so maybe I'm missing some reason it's better.

timj replied:
I used Generator because I looked at what you had done in another example. I can change.

TallJimbo commented on python/lsst/daf/butler/core/datastore.py:
```diff
@@ -667,12 +687,13 @@ def export(self, refs: Iterable[DatasetRef], *,
     @abstractmethod
-    def validateConfiguration(self, entities, logFailures=False):
+    def validateConfiguration(self, entities: Iterable[Union[DatasetRef, DatasetType, StorageClass]],
+                              logFailures: bool = False) -> None:
```
"Entity" might merit an alias.

TallJimbo commented on python/lsst/daf/butler/core/storageClass.py:
```diff
@@ -262,7 +267,7 @@ def filterParameters(self, parameters, subset=None):
         return {k: parameters[k] for k in known if k in parameters}
 
-    def validateInstance(self, instance):
+    def validateInstance(self, instance: object):
```
I have been using `typing.Any` instead of `object`; I _think_ they're equivalent.  We should probably standardize on one.

timj replied:
Yes, I'll use Any. What you are seeing is my experience with typing evolving over a week.

TallJimbo commented on python/lsst/daf/butler/core/storageClass.py:
```diff
@@ -278,14 +283,15 @@ def validateInstance(self, instance):
         return isinstance(instance, self.pytype)
 
-    def __eq__(self, other):
+    def __eq__(self, other: object):
```
Huh, I guess mypy doesn't complain about no return type annotation here?  I've been adding them for `__eq__`, but maybe I shouldn't be.

TallJimbo commented on python/lsst/daf/butler/core/storageClass.py:
```diff
@@ -353,7 +359,7 @@ class StorageClassFactory(metaclass=Singleton):
     """
 
-    def __init__(self, config=None):
+    def __init__(self, config: Optional[Union[StorageClassConfig, str]] = None):
```
Is `Union[StorageClassConfig, str, None]` better?  I'm not sure it is, but we should be intentional about establishing a convention one way or the other.

timj replied:
We've been using `Optional[something] = None` elsewhere so I'm fine with doing that.

TallJimbo commented on python/lsst/daf/butler/core/storageClass.py:
```diff
@@ -464,7 +470,7 @@ def processStorageClass(name, sconfig):
 
     @staticmethod
-    def makeNewStorageClass(name, baseClass=StorageClass, **kwargs):
+    def makeNewStorageClass(name, baseClass: Type = StorageClass, **kwargs) -> Type:
```
I think `Type[StorageClass]` would further check that arguments and return values are subclasses of `StorageClass`.

TallJimbo commented on python/lsst/daf/butler/datastores/chainedDatastore.py:
```diff
@@ -230,7 +248,7 @@ def exists(self, ref):
         return False
 
-    def get(self, ref, parameters=None):
+    def get(self, ref: DatasetRef, parameters: Mapping[str, Any] = None) -> Any:
```
Looks like this needs to be `Optional[Mapping[str, Any]]`.  Did mypy maybe not get enabled on this file?

timj replied:
It was enabled. I had convinced myself that all the parameter types where we assign a None default were automatically being tagged Optional by mypy (since clearly we are initializing with None). For that reason I stopped bothering with the explicit Optional. Maybe I was deluding myself.

TallJimbo replied:
Interesting!  Looks like you actually discovered a MyPy feature I didn't know about: https://mypy.readthedocs.io/en/stable/command_line.html#cmdoption-mypy-no-implicit-optional

But it sounds like disabling this behavior may become the default in the future, so we might want to be explicit with `Optional` anyway (and add that flag to our config).

timj replied:
Ok. I will make it explicit but it seems weird to me since it's so clear that I clearly mean Optional from the defaulting.

TallJimbo commented on python/lsst/daf/butler/datastores/chainedDatastore.py:
```diff
@@ -377,12 +395,13 @@ def isDatasetAcceptable(dataset, *, name, constraints):
         # NotImplementedError being raised.
         allFailuresAreNotImplementedError = True
         for datastore, constraints in zip(self.datastores, self.datastoreConstraints):
+            okForChild: List[FileDataset]
```
Ah, neat - I didn't know this syntax was possible.  I'm sure I'll use it in the future.

TallJimbo commented on python/lsst/daf/butler/datastores/fileLikeDatastore.py:
```diff
@@ -307,10 +325,13 @@ def addStoredItemInfo(self, refs, infos):
         self._table.insert(*records)
 
-    def getStoredItemInfo(self, ref):
+    def getStoredItemInfo(self, ref: Union[DatasetRef, FakeDatasetRef]) -> StoredFileInfo:
```
`DatasetIdRef` (in `registry.interfaces`) is a type alias for this union.

Open to ideas for better names for that.

TallJimbo commented on python/lsst/daf/butler/datastores/fileLikeDatastore.py:
```diff
@@ -799,7 +863,7 @@ def getUri(self, ref, predict=False):
         return location.uri
 
-    def get(self, ref, parameters=None):
+    def get(self, ref: DatasetRef, parameters: Mapping[str, Any] = None) -> Any:
```
Another one that should be `Optional`.

TallJimbo commented on python/lsst/daf/butler/datastores/posixDatastore.py:
```diff
@@ -19,22 +19,34 @@
+                    Optional,
+                    Type,
+                    Union
+                    )
```
Different formatting here than in the other files.

TallJimbo commented on python/lsst/daf/butler/core/datastore.py:
```diff
@@ -30,7 +30,19 @@
+                    Set,
+                    Type,
+                    Union,
+                    )
```
Another differently-formatted import pattern.

TallJimbo commented on python/mypy.ini:
```diff
@@ -33,3 +33,12 @@ disallow_incomplete_defs = True
+disallow_incomplete_defs = True
+strict_equality = True
+warn_unreachable = True
+warn_unused_ignores = True
```
Can you also turn mypy on for a few `core` modules after the changes you've made to them?  Or does that just bring new errors?

TallJimbo reviewed (approved):
Lots of very minor comments, none of them critical.  Overall this is of course a very nice cleanup.

timj committed:
Add allRefs method to DatasetRef and FakeDatasetRef

This simplifies the case where we only want the
refs for a single ref. Also makes type annotations
easier.

timj committed:
Rewrite StoredFileInfo with __init__ and frozen

Rather than using __post_init__ (which is very hard
to use type annotations with) instead force the
dataclass to be frozen but also replace with
our own __init__

timj committed:
Replace Datastore.names with a read-only tuple

A mutable list is not what we wanted for something
that should not be changed by the caller.

timj committed:
Enable type annotations for datastore

Also added annotations for related classes called
by datastores.

Mypy has been enabled for some core classes.

timj committed:
Use DatasetIdRef type rather than explicit FakeDatasetRef union
